pkg/agent/iptables/iptables.go

```diff
@@ -118,6 +162,130 @@ func (c *Client) SetupRules() error {
 		if err := c.ensureRule(rule.table, rule.chain, ruleSpec); err != nil {
 			return err
 		}
+		c.storeRule(rule.table, rule.chain, ruleSpec)
```

Comment:
why we need to store them in-memory? Didn't see the cache are consumed after initialization.

Reply:
because iptables changes between different mode, we do reconciliation as agent start. 
The logic is as follows:

1. save all rules installed so far
2. after route_controller finishes process all the known nodes during its own reconcile stage, it calls iptable.Reconcile.
3. iptable.Reconcile retrieves all iptable entries that refers to and is in Antrea-xxx chain, if any of the rule does not match that in 1) is removed.

Reply:
I see the problem now, it should still not work, as route_controller only remove stale routes and ports but not install required iptables, `iptable.Reconcile` will still remove nodes' rules, executing the Reconcile in route_controller has no difference with executing inside `iptables.SetupRules`.
And do we miss the logic to remove iptables rules for nodes in node deletion handling?

Reply:
This is what happens with regards to iptable and route table reconciliation, hope it helps,

1. route.Reconcile logic removes all stale routes from last incarnation
2.iptable.Reconcile logic removes all stale iptable rules from last instance. Including those rules setup in iptables.SetupRules in previous instance, and are no longer needed now.
3. We don't have logic to iptable rule deletion when node delete because all CIDRs present must be part of cluster CIDR.

Reply:
1. Between `iptables.SetupRules` and `iptables.Reconcile`, there are no other iptables rules (for node CIDRs) are installed,`iptables.Reconcile` will still remove previously installed and still required existing rules, and even if it's expected, there's no point to put `iptables.Reconcile` in route_controller as the cleanup can just be executed in `iptables.SetupRules`. Note that `removeStaleGatewayRoutes` and `removeStaleTunnelPorts` won't remove routes and ports that belong to existing nodes.
2. That would lead to orphaned iptables rule, though harmless, but the same applies to almost every resources including routes, openflows, container interfaces, aren't orphaned resources a bug?
@jianjuns @antoninbas

Reply:
The problem was that iptable.reconcile may be called before nodes had the opportunity to add per podCIDR rules. 

The solution is that before calling iptable.reconcile, node_route_controller iterate through all known nodes, and add per podCIDR rule for each of them.